---
title: "Differential Expression Analyses"
date: "`r format(Sys.time(), '%d %B, %Y')`"

output:
  bookdown::html_document2:
    number_sections: false
    global_numbering: true
    theme: spacelab
    highlight: pygments
    code_folding: show
    toc: true
    toc_float: true
    toc_depth: 2
    df_print: paged
---


# Setup

```{r setup}
# .libPaths(new = "/scratch/local/rseurat/pkg-lib-4.1.3")

suppressMessages({
  library(magrittr)
  library(tidyverse)
  library(Seurat)
})


set.seed(8211673)

knitr::opts_chunk$set(echo = TRUE, format = TRUE, out.width = "100%")


options(parallelly.fork.enable = FALSE,
        future.globals.maxSize = 8 * 1024 ^ 2 * 1000)

plan("multicore", workers = 8)
```

```{r useful-information, echo=FALSE}
cat("work directory: ", getwd())
cat("\n")
cat("library path(s): ", .libPaths())
```

# Load Data

We'll be working with the data from Day 1 ("First steps"), let's quickly re-load and re-process again:

```{r initial, warning=FALSE}
pbmc <- Read10X(data.dir = "./datasets/filtered_gene_bc_matrices/hg19/") %>%
  CreateSeuratObject(counts = ., project = "pbmc3k", min.cells = 3, min.features = 200)

pbmc[["percent.mt"]] <- PercentageFeatureSet(pbmc, pattern = "^MT-")

pbmc <- subset(pbmc, subset = nFeature_RNA > 200 & nFeature_RNA < 2500 & percent.mt < 5)

pbmc <- NormalizeData(pbmc, verbose = FALSE)

pbmc <- FindVariableFeatures(pbmc, verbose = FALSE)

pbmc <- ScaleData(pbmc, features = rownames(pbmc), verbose = FALSE)

pbmc <- RunPCA(pbmc, features = VariableFeatures(pbmc), verbose = FALSE)

pbmc <- FindNeighbors(pbmc, dims = seq_len(10), verbose = FALSE)

pbmc <- FindClusters(pbmc, resolution = 0.5, verbose = FALSE)

pbmc <- RunUMAP(pbmc, dims = seq_len(10), verbose = FALSE)
```

# Searching for Markers
Having identified (and visualized) clusters of cells we want to learn about the most important genes and differential gene sets that may explain the different cluster solutions in molecular terms.

Probably you already have prior knowledge and genes of interest, but here we will focus on how to derive **marker genes** for any given groups of cells.

To this end, Seurat has different convenience functions for **Differential Expression (DE)** tests. 
The main function is `FindMarkers()`. Again there are many parameters, the main ones are:

- ident.1, ident.2: typically these denote specific cluster numbers ("identities") of cells, but any other annotation from the metadata will work as well.
- test.use: select from a wide range of statistical models and/ or methodologies (we end to use "MAST")
- ?FindMarkers: have a look at other parameters and their defaults when a Seurat objec is passed to this function

[TM: I don't understand the following paragraph, please clarify. One should highlight the relation to DESEQ, NB, dispersion estimates. 
I guess it is a) easier because a cluster tends to have many more cells, and b) harder because of sparsity. Not sure how MAST fits into this.]

We'll be using `MAST`, because it accounts for the bimodality of single-cell data by jointly modeling rates of expression (discrete) and positive mean expression (continuous) values.Information from the discrete and continuous parts is combined to infer changes in expression levels using gene or gene set-based statistics. Given that this approach uses a generalized linear framework, it can be used to jointly estimate nuisance variation from biological and technical sources, as well as biological effects of interest.

**In a nutshell: MAST controls for the proportion of genes detected in each cell,** this acts as a proxy for both technical (e.g., dropout, amplification efficiency) and biological factors (e.g., cell volume and extrinsic factors other than treatment of interest) that globally influence gene expression.

<!-- https://www.bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#recommendations-for-single-cell-analysis

This research article that compares different DE testing methodologies/ package libraries for single-cell datasets: <https://www.nature.com/articles/nmeth.4612>
-->

Let's do a comparison between cells that were assigned to either cluster 1 or 2.

```{r findmarkers}
markers1v2 <- FindMarkers(
  object = pbmc,
  test.use = "MAST",
  ident.1 = 1,
  ident.2 = 2,
  verbose = FALSE
)
```

The resulting object is a data frame with the following columns:

- `p_val`: p-value (unadjusted!)
- `avg_log2FC` : log fold-change of the average expression between the two groups. Positive values indicate that the feature is more highly expressed in the first group.
- `pct.1`: The percentage of cells where the feature is detected in the first group
- `pct.2`: The percentage of cells where the feature is detected in the second group
- `p_val_adj`: Adjusted p-value, based on Bonferroni correction using all features in the dataset.

You may inspect the results with either `View()` or `head()`.

[TM: I remove the DT reference because this is not a course on markdown formatting.]

**Poll** What is the marker gene with the largest upregulation in cluster 1?

> âŒ¨ðŸ”¥ Exercise: What would happen if we used `ident.1 = 2`, and `ident.2 = 1` instead?


**Notice**: 

- if `ident.2=NULL` (default), then `FindMarkers()` will run a test between the group `ident.1` and **all other** cells
- You may also use a vector (e.g. `c(1,3)`) as `ident.2` to compare against all the cells of clusters one and three, pooled together.
- to increase the speed and relevance of marker discovery, Seurat allows for pre-filtering of features or cells. For example, genes that are very infrequently detected in either group of cells, or genes that are expressed at similar average levels, are unlikely to be differentially expressed, so we can exclude those: `?FindClusters` 

There are different flavors of these `Find*` functions. **It's important to get to know them!**

- `FindMarkers` will find markers between two different identity groups - you have to specify both identity groups. **This is useful for comparing the differences between two specific groups.**

- `FindAllMarkers` will find markers differentially expressed in each identity group by comparing it to all of the others - you don't have to manually define anything. Note that markers may bleed over between closely-related groups - they are not forced to be specific to only one group. 

- `FindConservedMarkers` will find markers that are conserved between two groups - this can be useful if you want to find markers that are conserved between a treated and untreated condition, for a specific cell type or group of cells. **It means they are differentially expressed compared to other groups, but have similar expression between the two groups you're actually comparing.**


## Graphical Exploration

Seurat includes several tools for visualizing marker expression. `VlnPlot()` and `RidgePlot` show the probability density distributions of expression values across clusters, and `FeaturePlot()` visualizes feature expression in a 2D projection (ee.g. PCA, t-SNE, or UMAP). 

[TM: do we need CellScatter latter? Otherwise I would skip it here.]

```{r markers_viz}
feats = rownames(markers1v2)[1:2]    # define genes of interest

VlnPlot(pbmc, features = feats)      # violin alternative: RidgePlot(pbmc, features = feats) 

FeaturePlot(pbmc, features = feats)  # the equivalent of DimPlot() for quantitative features 
```


## Markers Per Cluster (between)

What are the differences between clusters? disregarding any other metadata (cell-cycle, treatments, groups, etc). Let's compare 1 vs all, 2 vs all, ... etc.

For computational efficiency we'll use `pbmc_small` which is a downsampled PBMC3K data set bundled with Seurat. (Yes, it won't appear in your Global Environment but it's ready to be loaded upon execution of code.)

```{r allMarkers_pbmcsmall}
markers.between.clusters <- FindAllMarkers(
  pbmc_small,
  test.use = "MAST",
  logfc.threshold = 0.125,
  min.pct = 0.05,
  only.pos = TRUE,
  densify = TRUE
)

print(markers.between.clusters)
```

## DotPlot

[TM: Still need some text and explanation here]

```{r dotplot}
# notice that we take the markers from pbmc_small, but use them tto dotplot pbmc
DotPlot(pbmc, features = rownames(markers.between.clusters)[1:10])
```



## Optional
[TM: since the able formatting is mostly for enrichr (rmd/23) I would deal with it there.
I tthink at this stage we should have enoough manual exploration of marker genes, so the corresponding exercise can also be removed]
It is not uncommon that 
We'd like to rearrange columns, adjust p-values with a more sensitive approach, convert cluster column (if there is one) to character instead of factor, etc. We'll have all these tasks encapsulated into a function that should process the results from any of Seurat's `Find*()` functions.

```{r}
process_markers <- function(markers) {
  # first, let's get a boolean vector with values 'T'
  # on the positions where p_val is misrepresented as a zero
  nulls <- markers$p_val == 0
  if (any(nulls)) {
    # assign to these, the minimum positive value we can represent
    markers[nulls, ]$p_val <- .Machine$double.xmin
  }
  # re-run adjustment with default values
  markers$Adjusted.P.value <- p.adjust(markers$p_val)
  # run an assertion, it shouldn't hurt
  stopifnot(all(markers$Adjusted.P.value != 0))
  # add a column with gene names, just in case
  if (not("gene" %in% colnames(markers))) {
    markers$gene <- rownames(markers)
  }
  # filter non significant rows
  markers %<>% filter(Adjusted.P.value < 5e-2)
  # de-factor our cluster assignment, if there is one
  if ("cluster" %in% colnames(markers)) {
    markers$cluster <- as.numeric(as.character(markers$cluster))
    markers %<>% arrange(Adjusted.P.value) %>% select(
      gene, cluster, Adjusted.P.value, avg_log2FC, pct.1, pct.2
    )
  } else {
    markers %<>% arrange(Adjusted.P.value) %>% select(
      gene, Adjusted.P.value, avg_log2FC, pct.1, pct.2
    )
  }
  # in either case, we run a re-arrangement of rows.
  markers
}
```

```{r}
if ("Adjusted.P.value" %in% names(markers.between.clusters)) {
  print("Already processed.")
} else {
  markers.between.clusters %<>% process_markers()
}
```

Since `process_markers` is not **idempontent**, running the above codeblock would be troublesome if it weren't for the conditional (`if-else`) in place. Since we may sometime re-run cells as part of using the button left to the green triangle at each block... this structure can save us some headaches.


```{r}
DT::datatable(markers.between.clusters, rownames=NULL)
```

# End

[TM: I would have a similar exercise earlier. Also did not understand the "guessing" of celltypes. But I agree that wee could a few more test/polls/exercies above.]

> âŒ¨ðŸ”¥ Exercise: We'll distribute the room in groups again (N<=9), and assign a cluster to each. Get the "gene signature" of your assigned cluster, see if you can guess the celltype by using the following list of gene markers.
> 
> ```{r}
> list(
>   `Naive CD4+ T` = c("IL7R", "CCR7"),
>   `CD14+ Mono` = c("CD14", "LYZ"),
>   `Memory CD4+` = c("IL7R", "S100A4"),
>   B = c("MS4A1"),
>   `CD8+ T` = c("CD8A"),
>   `FCGR3A+ Mono` = c("FCGR3A", "MS4A7"),
>   NK = c("GNLY", "NKG7"),
>   DC = c("FCER1A", "CST3"),
>   Platelet = c("PPBP")
> )
> ```

Apart from getting and visualizing marker genes, we would ulimately want to seee whether there are common functional themes in those marker lists. This will be covered in the next section.
