---
title: "Differential Expression Analyses"
date: "`r format(Sys.time(), '%d %B, %Y')`"

output:
  bookdown::html_document2:
    number_sections: false
    global_numbering: true
    theme: spacelab
    highlight: pygments
    code_folding: show
    toc: true
    toc_float: true
    toc_depth: 2
    df_print: paged

params:
  Nthreads:
    label: 'Number of Threads:'
    value: 8
    input: slider
    min: 4
    max: 96
  Nmemgb:
    label: 'Gigabytes of Memory per Thread:'
    value: 8
    input: slider
    min: 4
    max: 16
  fork:
    label: 'Fork processes (not GUI)'
    value: FALSE
    input: checkbox
  seed:
    label: 'Random seed:'
    value: 8211673
  dimensionality:
    label: 'Dimensionality of single-cell dataset:'
    value: 10
  cluster_resolution:
    label: 'Parameter for clustering algorithm:'
    value: 0.5
---


# Setup

```{r}
cbind(params)
```

```{r}
library(tidyverse)
library(magrittr)
library(Seurat)
```

```{r}
set.seed(params$seed)
knitr::opts_chunk$set(echo = TRUE, format = TRUE, out.width = "100%")

options(
  parallelly.fork.enable = params$fork,
  future.globals.maxSize = params$Nmemgb * 1024^2 * 1000
)

plan("multicore", workers = params$Nthreads)
```

## Load Data

We'll be working with the data from Day 1 ("First steps"), let's quickly re-load and re-process again:

```{r}
pbmc <- Read10X(data.dir = "./datasets/filtered_gene_bc_matrices/hg19/") %>%
  CreateSeuratObject(counts = ., project = "pbmc3k", min.cells = 3, min.features = 200)

pbmc[["percent.mt"]] <- PercentageFeatureSet(pbmc, pattern = "^MT-")

pbmc <- subset(pbmc, subset = nFeature_RNA > 200 & nFeature_RNA < 2500 & percent.mt < 5)

pbmc <- NormalizeData(pbmc, verbose = FALSE)

pbmc <- FindVariableFeatures(pbmc, verbose = FALSE)

pbmc <- ScaleData(pbmc, features = rownames(pbmc), verbose = FALSE)

pbmc <- RunPCA(pbmc, features = VariableFeatures(pbmc), verbose = FALSE)

pbmc <- FindNeighbors(pbmc, dims = seq_len(params$dimensionality), verbose = FALSE)

pbmc <- FindClusters(pbmc, resolution = params$cluster_resolution, verbose = FALSE)

pbmc <- RunUMAP(pbmc, dims = seq_len(params$dimensionality), verbose = FALSE)
```

# Searching for Markers

Seurat has different functions that do all Differential Expression (DE) tests for you. The main one is `FindMarkers()`. Between other arguments, it accepts "identities" to establish --via metadata-- which groups of cells you're comparing. Another important argument is `test.use`, which allows you to select from a wide range of statistical models and/ or methodologies. We'll be using `MAST`, because it accounts for the bimodality of single-cell data by jointly modeling rates of expression (discrete) and positive mean expression (continuous) values. Information from the discrete and continuous parts is combined to infer changes in expression levels using gene or gene set-based statistics. Given that this approach uses a generalized linear framework, it can be used to jointly estimate nuisance variation from biological and technical sources, as well as biological effects of interest.

**In a nutshell: MAST controls for the proportion of genes detected in each cell,** this acts as a proxy for both technical (e.g., dropout, amplification efficiency) and biological factors (e.g., cell volume and extrinsic factors other than treatment of interest) that globally influence gene expression.

<!-- https://www.bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#recommendations-for-single-cell-analysis -->

Let's do a comparison between cells that were assigned to either cluster 1 or 2.

```{r}
markers1v2 <- FindMarkers(
  object = pbmc,
  test.use = "MAST",
  ident.1 = 1,
  ident.2 = 2,
  verbose = FALSE
)
```

The resulting object is a data frame with the following columns:

- `p_val`: p-value (unadjusted!)
- `avg_log2FC` : log fold-change of the average expression between the two groups. Positive values indicate that the feature is more highly expressed in the first group.
- `pct.1`: The percentage of cells where the feature is detected in the first group
- `pct.2`: The percentage of cells where the feature is detected in the second group
- `p_val_adj`: Adjusted p-value, based on Bonferroni correction using all features in the dataset.

You may inspect the results with either `View()` or `DT::datatable()`, according to your situation (e.g. working interactively at RStudio IDE or, rendering an HTML Report).

> âŒ¨ðŸ”¥ Exercise: What would happen if we used `ident.1 = 2`, and `ident.2 = 1` instead?

**If the `ident.2` parameter is omitted or set to `NULL`, `FindMarkers()` will test for differentially expressed features between the group specified by `ident.1` and all other cells.** You may also use a vector (e.g. `c(1,3)`) as `ident.2` to compare against all the cells of clusters one and three, pooled together.

To increase the speed of marker discovery, particularly for large datasets, Seurat allows for pre-filtering of features or cells. For example, features that are very infrequently detected in either group of cells, or features that are expressed at similar average levels, are unlikely to be differentially expressed.

There are different flavors of these `Find*` functions. **It's important to get to know them!**

- `FindMarkers` will find markers between two different identity groups - you have to specify both identity groups. **This is useful for comparing the differences between two specific groups.**

- `FindAllMarkers` will find markers differentially expressed in each identity group by comparing it to all of the others - you don't have to manually define anything. Note that markers may bleed over between closely-related groups - they are not forced to be specific to only one group. 

- `FindConservedMarkers` will find markers that are conserved between two groups - this can be useful if you want to find markers that are conserved between a treated and untreated condition, for a specific cell type or group of cells. **It means they are differentially expressed compared to other groups, but have similar expression between the two groups you're actually comparing.**

## Graphical Exploration

Seurat includes several tools for visualizing marker expression. `VlnPlot()` shows probability density distributions of expression values across clusters, and `FeaturePlot()` visualizes feature expression on a tSNE or PCA plot. There's also `RidgePlot()`, `CellScatter()`, and `DotPlot()` as additional methods to view your dataset.

```{r}
VlnPlot(pbmc, features = c("MS4A1", "CD79A"))
```

```{r}
FeaturePlot(
  pbmc,
  features = c(
    "MS4A1",
    "GNLY",
    "CD3E",
    "CD14",
    "FCER1A",
    "FCGR3A",
    "LYZ",
    "PPBP",
    "CD8A"
  )
)
```

Usually, we'd be getting top DE genes to draw these plots, and to drive GSEA as a downstream analysis to ultimately draw our conclusions.

# Further comparisons

### Markers Per Cluster (between)

What are the differences between clusters? disregarding any metadata (cell-cycle, treatments, groups, etc). Let's compare 1 vs all, 2 vs all, ... etc.

In this case, we'll use `pbmc_small` which is a down sampled PBMC3K data set bundled with Seurat. (Yes, it won't appear in your Global Environment but it's ready to be loaded upon execution of code.)

```{r}
markers.between.clusters <- FindAllMarkers(
  pbmc_small,
  test.use = "MAST",
  logfc.threshold = 0.125,
  min.pct = 0.05,
  only.pos = TRUE,
  densify = TRUE
)
```

```{r}
print(markers.between.clusters)
```

We'd like to rearrange columns, adjust p-values with a more sensitive approach, convert cluster column (if there is one) to character instead of factor, etc.

```{r}
process_markers <- function(markers) {
  # first, let's get a boolean vector with values 'T'
  # on the positions where p_val is misrepresented as a zero
  nulls <- markers$p_val == 0
  if (any(nulls)) {
    # assign to these, the minimum positive value we can represent
    markers[nulls, ]$p_val <- .Machine$double.xmin
  }
  # re-run adjustment with default values
  markers$Adjusted.P.value <- p.adjust(markers$p_val)
  # run an assertion, it shouldn't hurt
  stopifnot(all(markers$Adjusted.P.value != 0))
  # add a column with gene names, just in case
  if (not("gene" %in% colnames(markers))) {
    markers$gene <- rownames(markers)
  }
  # filter non significant rows
  markers %<>% filter(Adjusted.P.value < 5e-2)
  # de-factor our cluster assignment, if there is one
  if ("cluster" %in% colnames(markers)) {
    markers$cluster <- as.numeric(as.character(markers$cluster))
    markers %<>% arrange(Adjusted.P.value) %>% select(
      gene, cluster, Adjusted.P.value, avg_log2FC, pct.1, pct.2
    )
  } else {
    markers %<>% arrange(Adjusted.P.value) %>% select(
      gene, Adjusted.P.value, avg_log2FC, pct.1, pct.2
    )
  }
  # in either case, we run a re-arrangement of rows.
  markers
}
```

```{r}
if ("Adjusted.P.value" %in% names(markers.between.clusters)) {
  print("Already processed.")
} else {
  markers.between.clusters %<>% process_markers()
}
```

Since `process_markers` is not **idempontent**, running the above codeblock would be troublesome if it weren't for the conditional (`if-else`) in place. Since we may sometime re-run cells as part of using the button left to the green triangle at each block... this structure can save us some headaches.

# End

> âŒ¨ðŸ”¥ Exercise(s):

> 0. We'll distribute the room in groups again (N<=9), and assign a cluster to each. Get the "gene signature" of your assigned cluster, see if you can guess the celltype by using the following list of gene markers.
> 
> ```{r}
> list(
>   `Naive CD4+ T` = c("IL7R", "CCR7"),
>   `CD14+ Mono` = c("CD14", "LYZ"),
>   `Memory CD4+` = c("IL7R", "S100A4"),
>   B = c("MS4A1"),
>   `CD8+ T` = c("CD8A"),
>   `FCGR3A+ Mono` = c("FCGR3A", "MS4A7"),
>   NK = c("GNLY", "NKG7"),
>   DC = c("FCER1A", "CST3"),
>   Platelet = c("PPBP")
> )
> ```

> 1. Use `str()` over three different syntaxes of accessors to meta data data frame columns: `pbmc@meta.data[["orig.ident"]]`, `pbmc@meta.data["orig.ident"]`, and `pbmc@meta.data$orig.ident`; Which one of these is useless if we were to add a new column (granted, replacing `orig.ident` with our column name)? Could we just use `cbind`, how would that be?

> 2. Add a fake 'treatment' metadata column using `sample(x=c("KO","WT"))` and evaluate a pair of 'Within Clusters' code blocks in the same way as we did at 'Per Cluster (between)' but using `markers_within_clusters` function.


> 3. Explore the number of cells from each treatment level in each cluster? Do we have underrepresented classes in any of these? Would it affect our downstream DE analyses? Hint: use a code block with this one-liner: `DT::datatable(pbmc@meta.data %>% dplyr::count(seurat_clusters, treatment))`


> 4. Have a look at this research article that compares different DE testing methodologies/ package libraries for single-cell datasets: <https://www.nature.com/articles/nmeth.4612>
