---
title: "Analysis of Single-Cell Transcriptomes"
author: "Bioinfo-Core @ MPI-IE"
output:
  html_document:
    toc: false
---

In this course we will use R and the Seurat package to introduce the basic analysis steps and quality controls for single-cell RNA data. The target audience are PhD students and postdocs who have previous experience with R language.

> Seurat is an R package designed for QC, analysis, and exploration of
> single-cell RNA-seq data. Seurat aims to enable users to identify and
> interpret sources of heterogeneity from single-cell transcriptomic
> measurements, and to integrate diverse types of single-cell data. All methods
> emphasize clear, attractive, and interpretable visualizations. Seurat is
> developed and maintained by the Satija lab and is released under the MIT
> license.

**Goal:** After the course, you will understand the challenges of single-cell RNA datasets, be able to conduct quality control of such data, and use Seurat to conduct your own analysis.


### Outlook

- Single-cell sequencing protocols
- Importing 10X Chromium Data
- Quality Control
- Running Seurat
- Normalization, and Scaling Data
- Dimensional Reduction
- Clustering Cells with their Neighborhoods
- Identification of Marker genes
- Data transformations and Batch Effects correction
- Integration of multiple datasets
- Querying Cell Identities


### Requirements

This is a hands-on course and requires **(1)** a WLAN-capable laptop with Internet access (your device needs to be registered with Core-IT). **(2)** An account for [Workbench](https://workbench.ie-freiburg.mpg.de). We will use the Rstudio server at the MPI-IE. If you do not have an account yet, you will need to request a 'linux' account from core-IT (please do so well before the course starts). We will be using R `4.1.3` and Bioconductor `3.14` during this course.

Even though a simple laptop may not be powerful enough to handle larger analysis, you can use a local installation of R as well. Download it from here for your platform: [Linux](https://cran.rstudio.com/bin/linux/), [Mac](https://cran.rstudio.com/bin/macosx/base/), [Windows](https://cran.rstudio.com/bin/windows/base/old/).

Some previous knowledge is assumed:

* **R-language**: has been covered by our introductory
  [course](https://maxplanck-ie.github.io/Rintro).

* **Bulk RNA-seq**: has been covered by our introductory
  [course](https://maxplanck-ie.github.io/Rdeseq2).

* **Your keyboard!** Special characters will be needed throughout the
  course, e.g. $ , | , [, ], {, }, >, <, #, ~, &, ^, %, !, ?


Course language: English

***
